MoreLinq/EquiZip.cs

```diff
@@ -152,38 +153,28 @@ static partial class MoreEnumerable
         }
 
         static IEnumerable<TResult> EquiZipImpl<T1, T2, T3, T4, TResult>(
-            IEnumerable<T1> first,
-            IEnumerable<T2> second,
-            IEnumerable<T3> third,
-            IEnumerable<T4> fourth,
+            IEnumerable<T1> s1,
+            IEnumerable<T2> s2,
+            IEnumerable<T3> s3,
+            IEnumerable<T4> s4,
             Func<T1, T2, T3, T4, TResult> resultSelector)
         {
-            using (var e1 = first.GetEnumerator())
-            using (var e2 = second.GetEnumerator())
-            using (var e3 = third?.GetEnumerator())
-            using (var e4 = fourth?.GetEnumerator())
+            var limit = new object[] { s1, s2, s3, s4 }.TakeWhile(s => s != null).Count() - 1;
+
+            return ZipImpl(s1, s2, s3, s4, resultSelector, limit, enumerators =>
+            {
+                int shortest = enumerators.Index().First(x => x.Value == null).Key;
+
+                return new InvalidOperationException(GetOrdinal(shortest + 1) + " sequence too short.");
+            });
+
+            string GetOrdinal(int i)
```

Comment:
Use an array?

```c#
static readonly string[] OrdinalNumbers =
{
    "First",
    "Second",
    "Third",
    "Fourth",
};
```

It will help when extending the arity of zipping operators.